### PR TITLE
feat: Add percent coverage option to stat

### PR DESCRIPTION
**Pull request by cademirch**
This adds the feature requested in #81 by @Jakob37. I tried my best to emulate the existing code in `stat`, but would of course appreciate any feedback. 

Not sure if you're accepting new features @brentp, happy to discuss more.

### CODE REVIEW

File: d4/src/task/perc_cov.rs [hunk 1]
```diff
@@ -0,0 +1,83 @@
...
+        for part in parts {
+            for (i, &value) in part.iter().enumerate() {
+                sums[i] += value;
+            }
+        }
+
+        let result: Vec<f32> = sums.into_iter()
+        .map(|x| {
+            if x != 0 {
+                (x - 1) as f32 / divisor
```
Comment (brentp) [inline d4/src/task/perc_cov.rs:74]: why are you subtracting 1 here? can you add a comment?

**Review by brentp — CHANGES_REQUESTED** [review]
this looks good to me, but would nice to start adding tests so we can catch regressions for future development.

**Comment by cademirch** [comment]
I agree tests are needed. @Jakob37 mentioned they can help with tests for this, and we can setup some actions to run them.

File: d4/src/task/perc_cov.rs (hunk 1 above)
Comment (cademirch) [reply d4/src/task/perc_cov.rs:74]: Iirc was off by 1 in my test case - I don't recall exactly why. I'll take a look at it to remind myself and update with a comment

Comment (cademirch) [reply d4/src/task/perc_cov.rs]: removed this as it was not needed. not sure what I was thinking then...

**Comment by Jakob37** [comment]
Great, thank you for putting together this PR.

> @Jakob37 could you possibly provide a small d4 file and regions.bed along with expected output?

I'll look into it!

**Comment by Jakob37** [comment]
I am a bit confused on how ranges are handled. When I run `d4tools show` it looks zero indexed, and it looks like the end of each position is the same as the start of the next. I.e. as how `ranges` are handled in Python.

In the bed file on the other hand, I would expect things to be one-indexed and having "inclusive" ends.

I.e. if having the ranges in the bed file `1 - 3` this would mean the first three positions. And if I want to show the same in Python, it would be `0 - 3`.

Is this correct?

Based on this, running just the two very narrow ranges (one indexed bed file):

```
19      45769557        45769559
```

And based on the coverage data (first lines in the provided d4 file, zero indexed with exclusive ends)

```
19	45769552	45769557	1
19	45769557	45769563	2
```

I would expect the coverage 0.667, but I get 1.

Also, on the second range (one indexed bed file):

```
19      45769557        45769562
```

I would expect 0.8 in my testing (i.e. 4 of them falling within the second range), but I get 1.4 when running the PR.

Have to run, I can see if I can clarify this tomorrow.

**Review by Jakob37 — COMMENTED** [review]
OK, I have set up a minimal test dataset with a 10bp reference and two reads. Here is relevant data (**edit:** with updated ranges): [pr_files_bedzero.zip](https://github.com/user-attachments/files/16392630/pr_files_bedzero.zip)

Illustration of the alignment and coverages:

```
   GGG
    GGAA
AAAGGGAACC
0001221100
```

The samtools depth profile (`samtools depth out.bam`) (1-indexed):

```
$ samtools depth out/out.bam
chr     4       1
chr     5       2
chr     6       2
chr     7       1
chr     8       1
```

Here are the four ranges in `ranges.bed`:

```
$ cat ranges.bed
chr	0	5
chr	3	5
chr	4	5
chr	5	10
```

From these I would expect the outputs (please double check):

```
#Chr    Start   End     1x      2x
chr     0       5       0.4     0.2
chr     3       5       0.5     1
chr     4       5       0       1
chr     5       10      0.6     0.2
```

Here is the current output:

```
$ <home>/src/d4-format-cademirch/target/debug/d4tools stat -H -s perc_cov=1,2 -r bwa_test/ranges_open.bed bwa_test/out/out.d4
#Chr    Start   End     1x      2x
chr     0       5       0.600   0.200
chr     3       5       1.500   0.500
chr     4       5       2.000   1.000
chr     5       10      0.800   0.400
```

I think this is really close, and very nice work so far. Results looks roughly correct when working with a bigger dataset, so I suspect this being one or two off-by-one errors.

**Edit: Indeed, one of the off-by-one was from my previous bed file where I mistakenly 1-indexed. But there is still something weird going on. Might it be related to  #75?**

I'll go on two weeks vacation now, but looking forward to helping bringing this into goal when back!

**Comment by cademirch** [comment]
Thank you for putting those datasets together and effort testing @Jakob37! I've always understood that bed coordinates should be 0 based and half open. 

I'm also going on vacation, but will be back on Tuesday and will take a look at this more then!

**Comment by Jakob37** [comment]
> Thank you for putting those datasets together and effort testing @Jakob37! I've always understood that bed coordinates should be 0 based and half open.

Aha, then I learned something. It looks like you are right. I'll update the range coordinates above then.

> I'm also going on vacation, but will be back on Tuesday and will take a look at this more then!

Sounds great. Enjoy your vacation!

**Comment by cademirch** [comment]
> Aha, then I learned something. It looks like you are right. I'll update the range coordinates above then.
0 vs 1 based genomic coordinates is the most annoying problem 🙄 

> Sounds great. Enjoy your vacation!
Thank you, you too!

**Comment by Jakob37** [comment]
I am back in office now. I am happy to help out further with testing / providing testing data @cademirch . Just let me know what you need. Excited to get this in, as it will mean that we can start using Chanjo2 in production 😄

**Comment by northwestwitch** [comment]
Hi @cademirch, I really appreciate this new feature! Good stuff!🥇 
Perhaps in this PR you could update the README file to include the new command before merging, so the changes reflect on the documentation?

**Comment by cademirch** [comment]
Hi @Jakob37 and @northwestwitch, will try to work on this later this week!

**Comment by cademirch** [comment]
There is some weird double counting happening in `feed_range`, with the ranges @Jakob37 made in the small dataset.

Added a print statement in `feed_range` to see whats happening:
```rust
fn feed_range(&mut self, left: u32, right: u32, value: &mut Once<i32>) -> bool {
        let value = value.next().unwrap();
        println!(
            "bed_region:({}, {}) feed_range:({}, {}), value:{:?}",
            self.region_begin, self.region_end, left, right, value
        );
        for (i, thresh) in self.thresholds.iter().enumerate() {
            if value as u32 >= *thresh {
                self.counts[i] += std::cmp::min(right, self.region_end) - left as u32
            }
        }
        true
    }
```

Output with `ranges.bed` unmodified from above:
```
❯ ./target/release/d4tools stat -H -s perc_cov=1,2 -r ../perc-cov-testing/ranges.bed ../perc-cov-testing/out.d4
bed_region:(0, 5) feed_range:(0, 3), value:0
bed_region:(0, 5) feed_range:(3, 3), value:1
bed_region:(0, 5) feed_range:(0, 3), value:0
bed_region:(0, 5) feed_range:(3, 4), value:1
bed_region:(3, 5) feed_range:(0, 3), value:0
bed_region:(3, 5) feed_range:(3, 4), value:1
bed_region:(0, 5) feed_range:(0, 3), value:0
bed_region:(0, 5) feed_range:(3, 4), value:1
bed_region:(3, 5) feed_range:(0, 3), value:0
bed_region:(3, 5) feed_range:(3, 4), value:1
bed_region:(4, 6) feed_range:(0, 3), value:0
bed_region:(4, 6) feed_range:(3, 4), value:1
bed_region:(0, 5) feed_range:(4, 5), value:2
bed_region:(3, 5) feed_range:(4, 5), value:2
bed_region:(4, 6) feed_range:(4, 5), value:2
bed_region:(4, 6) feed_range:(4, 6), value:2
#Chr    Start   End     1x      2x
chr     0       5       0.600   0.200
chr     3       5       1.500   0.500
chr     4       6       2.000   1.500
```

If we make `ranges.bed` just:
```
chr	0	5
chr	3	5
```

Output looks better:
```
❯ ./target/release/d4tools stat -H -s perc_cov=1,2 -r ../perc-cov-testing/ranges.bed ../perc-cov-testing/out.d4
bed_region:(0, 5) feed_range:(0, 3), value:0
bed_region:(0, 5) feed_range:(3, 3), value:1
bed_region:(0, 5) feed_range:(0, 3), value:0
bed_region:(0, 5) feed_range:(3, 4), value:1
bed_region:(3, 5) feed_range:(0, 3), value:0
bed_region:(3, 5) feed_range:(3, 4), value:1
bed_region:(0, 5) feed_range:(4, 5), value:2
bed_region:(3, 5) feed_range:(4, 5), value:2
#Chr    Start   End     1x      2x
chr     0       5       0.400   0.200
chr     3       5       1.000   0.500
```

**Comment by Jakob37** [comment]
> There is some weird double counting happening in feed_range.

Hmmm. Yes, there indeed seems to be something strange going on.

I tried reducing the ranges further, to get something minimal to debug:

```
chr 3   5
chr 4   5
```

This gave me:

```
$ /home/jakob/src/d4-format-cademirch/target/debug/d4tools stat -H -s perc_cov=1,2 -r ranges_minimal_debug.bed out.d4 
feed_rows 3 4
feed_range:(3, 4), value:1
feed_rows 3 4
feed_range:(3, 4), value:1
feed_rows 3 4
feed_range:(3, 4), value:1
feed_rows 4 5
feed_range:(4, 5), value:2
feed_rows 4 5
feed_range:(4, 5), value:2
#Chr    Start   End     1x      2x
chr     3       5       1.500   0.500
chr     4       5       2.000   1.000
```

As you say, it looks like there is some double counting going on. I'll poke around and see if I can learn some more about it.

**Comment by northwestwitch** [comment]
Hi! I gave a this branch a try using chanjo2 and as you say, looks like it's double the expected values.

This is for instance the values from chanjo2 main branch (completeness computed with an algorithm in chanjo2):

image

This is the result using this branch of d4tools:

image

It's not exactly double values because the stats are calculated in a different way, but kindof 🤔

**Comment by Jakob37** [comment]
I suspect something is going wrong in the `scan_partition_impl` call in the `track.rs` file.

`scan_partition_impl` is called multiple times (once per range), and `active_handles.iter_mut()` is also processing specific ranges multiple times (up to once per range).

https://github.com/38/d4-format/blob/68d49025c8a3e57e90486e6566e8f24a8e125f06/d4/src/d4file/track.rs#L181-L198

The issue seems to be triggered at the bounds between different coverage levels.

This is fine:

```
#Chr    Start   End     1x      2x
chr     2       5       0.667   0.333
chr     3       5       1.000   0.500
```

But if you add 1 to the 3-5 range:

```
#Chr    Start   End     1x      2x
chr     2       5       1.000   0.333
chr     4       5       2.000   1.000
```

If extending it to 2nts on the right side, it is still weird:

```
#Chr    Start   End     1x      2x
chr     2       5       1.000   0.333
chr     4       6       2.000   1.500
```

Can you guys see what might be going wrong here?

I'll circle back to this later this week, to see it again with fresh eyes.

**Comment by cademirch** [comment]
Thanks for taking a look at this @Jakob37 and @northwestwitch. I too ended up at the `scan_partition_impl` closure in my debugging. I think the issue stems from `last_right` being set to `part_left` at the beginning of each partition scan. When this happens with multiple bed ranges that span multiple partitions (I believe partitions are defined where coverage changes), then double feeds occur. My solution is to add  a `global_last_right` outside the closure to keep track the last right position across partitions:

```rust
fn scan_partition<DS: DataScanner<Self::RowType>>(&mut self, handles: &mut [DS]) {
        let default_primary_value = self.primary.default_value();
        let mut decoder = self.primary.make_decoder();
        let mut global_last_right = None; // Track the last right across all partitions
        let mut part_num = 0;
        scan_partition_impl(handles, |part_left, part_right, active_handles| {
            if let Some(default_value) = default_primary_value {
                part_num += 1;
                let iter = self.secondary.seek_iter(part_left);
                let mut last_right = global_last_right.unwrap_or(part_left);
                // let mut last_right = part_left; 
                println!("scan_partition {} part left: {} part right: {} last_right: {}, global: {:?}", part_num, part_left, part_right, last_right, global_last_right);
                for (mut left, mut right, value) in iter {
                    left = left.max(part_left);
                    right = right.min(part_right).max(left);
                    println!("scan_partition {} left: {} right: {}, last_right: {}", part_num, left, right, last_right);
    
                    if last_right >= left {
                        left = last_right;
                    }
...

Outputs: (Print debugging saves the day 😄)
```
❯ RUST_BACKTRACE=1 ./target/release/d4tools stat -H -s perc_cov=1,2 -r ../perc-cov-testing/ranges.bed ../perc-cov-testing/out.d4
scan_partition 1 part left: 0 part right: 3 last_right: 0, global: None
scan_partition 1 left: 3 right: 3, last_right: 0
feed_rows 1 last_right:0, left:3
feedrange bed_region:(0, 5) feed_range:(0, 3), value:0
feed_rows 1 left:3, right:3
feedrange bed_region:(0, 5) feed_range:(3, 3), value:1
scan_partition 2 part left: 0 part right: 4 last_right: 3, global: Some(3)
scan_partition 2 left: 3 right: 4, last_right: 3
feed_rows 2 left:3, right:4
feedrange bed_region:(0, 5) feed_range:(3, 4), value:1
feed_rows 2 left:3, right:4
feedrange bed_region:(3, 5) feed_range:(3, 4), value:1
scan_partition 3 part left: 0 part right: 5 last_right: 4, global: Some(4)
scan_partition 3 left: 3 right: 4, last_right: 4
feed_rows 3 left:4, right:4
feedrange bed_region:(0, 5) feed_range:(4, 4), value:1
feed_rows 3 left:4, right:4
feedrange bed_region:(3, 5) feed_range:(4, 4), value:1
feed_rows 3 left:4, right:4
feedrange bed_region:(4, 6) feed_range:(4, 4), value:1
scan_partition 3 left: 4 right: 5, last_right: 4
feed_rows 3 left:4, right:5
feedrange bed_region:(0, 5) feed_range:(4, 5), value:2
feed_rows 3 left:4, right:5
feedrange bed_region:(3, 5) feed_range:(4, 5), value:2
feed_rows 3 left:4, right:5
feedrange bed_region:(4, 6) feed_range:(4, 5), value:2
scan_partition 4 part left: 4 part right: 6 last_right: 5, global: Some(5)
scan_partition 4 left: 4 right: 6, last_right: 5
feed_rows 4 left:5, right:6
feedrange bed_region:(4, 6) feed_range:(5, 6), value:2
#Chr    Start   End     1x      2x
chr     0       5       0.400   0.200
chr     3       5       1.000   0.500
chr     4       6       1.000   1.000
```
Note for the bed range `chr 3 5` I think this is the correct output since base 3 has a depth of 1 and base 4 has a depth of 2, thus 100% of bases are >= 1x and 50% are >= 2x. 

I will setup up some more tests for this fix and make sure this doesn't break anything else. Should probably make a separate PR for this? Maybe @brentp has a preference?

**Comment by Jakob37** [comment]
Great job debugging @cademirch ! Home with sick kid today, but will test and see if I can "break" your new code later this week 😅

**Comment by cademirch** [comment]
My workaround seems to break the test in `d4tools/test/test-results/stat/region-stat`. Still trying to wrap my head around it. In the test, none of the bed regions are overlapping which seems to be the crux of the issue in this PR. It seems like `scan_partition_impl` was designed with the expectation that it would be serving only one bed interval some data (this is speculation). It may take some more clever solution to handle both cases.

**Comment by Jakob37** [comment]
I have done some more digging around. Takes some time to penetrate this!

I have another solution-candidate.

In this chunk:

https://github.com/38/d4-format/blob/68d49025c8a3e57e90486e6566e8f24a8e125f06/d4/src/d4file/track.rs#L66-L84

The active handles are overlapping ranges. 

`func` will retrieve values from the d4file within the range and perform the supplied stat calculation for each handle (`active_heap`).

I used a simple example with intervals `0 - 10` and `4 - 5`.

Here I could see that something went wrong. Calculations was performed for ranges:

1. For range `0 - 4`
2. For range `4 - 5`
3. For range `0 - 10` (this is the error - should be `5 - 10` I think)

Here are my debug prints. The `0 - 10` result is wrong - we expects 6 bases above 1x and 3 above 2x.

```
func (2) called for range: 0 4
  active_heap[0]: 0 10
func (1) called for range: 4 5 (last_end 4)
  active_heap[0]: 4 5
  active_heap[1]: 0 10
func (1) called for range: 0 10 (last_end 5)
  active_heap[0]: 0 10
     No more handles left
#Chr    Start   End     1x      2x
chr     0       10      0.800   0.400
chr     4       5       1.000   1.000
```

I updated it such that it only performs calculations for intervals once. I needed to update this both in line 73 and 86.

```
let this_begin;
if top.get_range().0 > last_end {
    this_begin = top.get_range().0;
} else {
    this_begin = last_end;
}
```

Then it calculates for the expected ranges, and gives the expected results.

```
func (2) called for range: 0 4
  active_heap[0]: 0 10
func (1) called for range: 4 5 (last_end 4)
  active_heap[0]: 4 5
  active_heap[1]: 0 10
func (1) called for range: 5 10 (last_end 5)
  active_heap[0]: 0 10
     No more handles left
#Chr    Start   End     1x      2x
chr     0       10      0.600   0.300
chr     4       5       1.000   1.000
```

Trying it out on the ranges you used:

```
func (2) called for range: 0 3
  active_heap[0]: 0 5
func (2) called for range: 3 4
  active_heap[0]: 0 5
  active_heap[1]: 3 5
func (1) called for range: 4 5 (last_end 4)
  active_heap[0]: 0 5
  active_heap[1]: 3 5
  active_heap[2]: 4 6
func (1) called for range: 5 6 (last_end 5)
  active_heap[0]: 4 6
#Chr    Start   End     1x      2x
chr     0       5       0.400   0.200
chr     3       5       1.000   0.500
chr     4       6       1.000   1.000
```

This might very well have been a follow-up issue from the previous "fix" I made.

I will do some further testing here as well, and see if I can run this in the full test suite.

**Comment by Jakob37** [comment]
Looks like the change above passes the tests (guessing the remove-view errors are unrelated).

tests

**Comment by cademirch** [comment]
Wow, awesome @Jakob37! I was planning on picking up where I left off on this tomorrow morning, but it looks like you've figured it out! And I think this is the best solution since this is fixing the issue in the `scan_partition_impl` definition, rather than the closure we were looking at before.

**Comment by Jakob37** [comment]
> Wow, awesome @Jakob37! I was planning on picking up where I left off on this tomorrow morning, but it looks like you've figured it out! And I think this is the best solution since this is fixing the issue in the `scan_partition_impl` definition, rather than the closure we were looking at before.

Yes, I agree.

But then it seems we can start wrapping this up by adding tests and updating the docs. 

I plan to test some more edge cases later today. Will let you know how it goes.

If there is something else I can help out with, let me know!

**Comment by cademirch** [comment]
Great. I'll do some testing as well. Will you make a new PR with your fix?  Then if all is good @brentp can merge that and then this one.

**Comment by Jakob37** [comment]
> Great. I'll do some testing as well. Will you make a new PR with your fix? Then if all is good @brentp can merge that and then this one.

OK, I'll put up the PR!

**Comment by Jakob37** [comment]
PR up: #84

**Comment by Jakob37** [comment]
Merged! Let's wrap this one up now. Seems like the remaining steps are:

- Document the changes in the command line interface
- Document the changes in the documentation
- Add one or several tests based on the 10nt d4-file (now present in the test data folder)

Just let me know if you want help with any of this @cademirch

**Comment by cademirch** [comment]
@Jakob37 great! I'll try to get this done asap.

**Comment by cademirch** [comment]
alright @Jakob37 and @brentp, I think this is ready to merge!

File: d4tools/src/stat/main.rs [hunk 2]
```diff
@@ -227,6 +227,51 @@ fn hist_stat(matches: ArgMatches) -> Result<(), Box<dyn std::error::Error>> {
...
+        for t in thresholds.iter() {
+            print!("\t{}x", t);
+        }
+        println!();
+        print_header = false;
+    }
+    for r in results.into_iter() {
+        print!("{}\t{}\t{}", r.chrom, r.begin, r.end);
+        for value in r.output.iter() {
+            print!("\t{:.3}", value);
```
Comment (brentp) [inline d4tools/src/stat/main.rs:269]: I'm fine with this, but just noting that sometimes people will do huge ranges then it will say, for example `0.000` when there are bases above a given threshold. Might we use `:3g` or something?

File: d4/src/task/perc_cov.rs [hunk 3]
```diff
@@ -0,0 +1,77 @@
...
+        }
+        true
+    }
+    // #[inline(always)]
+    fn feed_range(&mut self, left: u32, right: u32, value: &mut Once<i32>) -> bool {
+        let value = value.next().unwrap();
+        for (i, thresh) in self.thresholds.iter().enumerate() {
+            if value as u32 >= *thresh {
+                self.counts[i] += right - left as u32
+            }
```
Comment (brentp) [inline d4/src/task/perc_cov.rs:36]: since thresholds are sorted, you can break in else, but that might be a net negative in most common case where there are few thresholds

**Review by brentp — APPROVED** [review]
LGTM minor comments you can choose to ignore.

File: d4tools/src/stat/main.rs (hunk 2 above)
Comment (cademirch) [reply d4tools/src/stat/main.rs:269]: Makes sense to me, not familiar with `:3g` though and couldn't find anything about that. Could only find this on precision: https://doc.rust-lang.org/std/fmt/#precision

File: d4/src/task/perc_cov.rs (hunk 3 above)
Comment (cademirch) [reply d4/src/task/perc_cov.rs:36]: ah that makes sense to me. will add.

File: d4tools/src/stat/main.rs (hunk 2 above)
Comment (Jakob37) [reply d4tools/src/stat/main.rs:269]: I agree with the thought here, that information might be lost by truncating at three digits.

Seems the `g` formatting is present in the `printf` in C and bash, and automatically selects between precision and scientific notation https://www.gnu.org/software/libc/manual/html_node/Floating_002dPoint-Conversions.html. On a quick test it didn't seem to be supported in Rust though (I might be wrong).

What about using a sane default (for instance as you have now) and letting the user control the number of decimals / showing the full value through a flag?

(I am also fine with it as is though - eager to get this in quickly so we can get things going 😅 )

File: README.md [hunk 4]
```diff
@@ -155,7 +155,7 @@ FLAGS:
 OPTIONS:
     -r, --region <bed_file_path>      A bed file that describes the region we want to run the stat
     -s, --stat <stat_type>            The type of statistics we want to perform, by default average. You can specify
-                                      statistic methods: mean, median, hist, percentile=X% (If this is not speficied
+                                      statistic methods: perc_cov, mean, median, hist, percentile=X% (If this is not speficied
```
Comment (Jakob37) [inline README.md:158]: There is also a file `d4tools/src/stat/cli.yml` with this text. I think that controls what is shown in the `d4tools stat --help` text.

**Review by Jakob37 — APPROVED** [review]
Great job 🎉 

I did a test using a full size d4 file. Looks good!

```
$ ~/src/d4-format-cademirch/target/debug/d4tools stat --stat perc_cov=20,40,60 ~/data/hg002.d4 -r ~/data/d4_testing_regions/intervals.bed -H | head
#Chr    Start   End     20x     40x     60x
1       26051301        26068436        0.996   0.748   0.038
1       74235387        74544428        0.966   0.550   0.004
1       77888513        77943895        0.977   0.675   0.020
1       114704469       114716771       1.000   0.774   0.021
1       155897808       155911404       0.995   0.761   0.013
1       156082573       156140081       0.999   0.797   0.073
1       201359008       201377764       1.000   0.734   0.019
1       236664141       236764631       0.969   0.717   0.024
1       237042184       237833988       0.986   0.667   0.015
```

File: d4tools/src/stat/main.rs (hunk 2 above)
Comment (Jakob37) [reply d4tools/src/stat/main.rs:269]: Maybe something like this would work:

```
if value.abs() >= 0.001 {
    print!("\t{:.3}", value);
} else {
    print!("\t{:.3e}", value);
}
```

I.e. explicitly showing scientific notation for smaller numbers.

Comment (cademirch) [reply d4tools/src/stat/main.rs:269]: Good idea. Sorry for delay got caught up with teaching things as the quarter starts soon. Will get this squared away in the morning!

Comment (Jakob37) [reply d4tools/src/stat/main.rs:269]: No worries, many thanks 🙏

Comment (cademirch) [reply d4tools/src/stat/main.rs]: resolved in latest commits

File: d4/src/task/perc_cov.rs (hunk 3 above)
Comment (cademirch) [reply d4/src/task/perc_cov.rs:36]: changed my mind on this, i think you're right that the most common case will be few thresholds

**Comment by cademirch** [comment]
Alright I think this should be good to go. LMK if there is anything else!

**Comment by Jakob37** [comment]
Nice! If @brentp doesn't have further comments, then let's get it in 👌

**Comment by Jakob37** [comment]
Would you be able to merge and release @brentp ? Many thanks 🙏
